language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -27,9 +30,11 @@ import           PlutusPrelude
 -- builtin names.
 data BuiltinTable = BuiltinTable (M.Map TypeBuiltin (Kind ())) (M.Map BuiltinName (NormalizedType TyNameWithKind ()))
 
+type TypeSt = IM.IntMap (NormalizedType TyNameWithKind ())
```

Comment:
[Control.Lens.Zoom](https://hackage.haskell.org/package/lens-4.16.1/docs/Control-Lens-Zoom.html#t:Zoomed) provides a very nice way of dealing with nested state and functions that touch different parts of state. We could use it here, but `TypeSt` is very slightly different from `CloneSt`. Can we require cloning to be defined over normalized types?

Reply:
`CloneSt` must be locally unique, so I don't think we would want to do that. Requiring cloning be defined in the same way as `TypeSt` would not be good, as we are simply adjusting names while cloning; during type synthesis we perform substitutions that take names to arbitrary types.